accepted/future-releases/nnbd/feature-specification.md

```diff
@@ -616,6 +609,341 @@ It is a warning to use a null aware operator (`?.`, `?[]`, `?..`, `??`, `??=`, o
+inferred type of the variable shall be `X`.  However, such a variable shall be
+treated as immediately promoted to `X & T`.
+
+#### Untyped variables
+
+Local variables with no explicitly written type and no initializer are subject
+to further promotion based inference.  The variables under consideration here
+are those declared as `var x;`, `final x;`, `late var x;`, or `late final x;`.
+We refer to such variables as **untyped variables**.
+
+An **initializing assignment** is an assignment to an **untyped variable** when
+that variable is in the **definitely unassigned** state.  Such an assignment on
+a path shall cause the variable in question to be treated by the flow analysis
+on that path as if it were declared with the assigned expression as its
+initializing expression, up until the next join point in the program.  We refer
+to this inferred type as the **initialization inferred type** of the variable.
+The **initialization inferred type** of the variable is always treated as a
+**type of interest** for the purposes of promotion (including its non-nullable
+version if applicable).
+
+```dart
+void test() {
+  var x; // x is an untyped variable
+  x = 3; // x is given the initialization inferred type of `int`
+}
+```
+
+At a join point in the program, it is an error if an **untyped variable** has
+been given an **initialization inferred type** on one of the reachable paths to
+the join point and not on another reachable path, unless the variable is
+declared as `late`, or using `var`.  Note that it is valid for a variable to
+have no **initialization inferred type** on a path into a join point when the
+join point is known to be unreachable along that path.
+
+```dart
+void test(bool b) {
+  final x; // x is an untyped variable
+  if (b) {
+    x = 3;
+  } // Error, x has not been given an inferred type on all paths
+
+  late var y; // y is a late untyped variable
+  if (b) {
+    y = 3;
+  } // Ok, y has inferred type `int`
+
+  final z;
+  if (b) {
+     z = 3;
+  } else {
+    return;
+  }  // Ok, z has inferred type `int` even though it is unassigned on one path.
+}
+```
+
+At a join point in the program, if an **untyped variable** has been given an
+**initialization inferred type** on all of the reachable paths into the join and
+at least one of the paths is reachable, then the **initialization inferred
+type** for the variable after the join point is computed as follows:
+  - If the **initialization inferred type** on each reachable path is the same
+    type `T` (using syntactic equality), then the **initialization inferred
+    type** after the join is `T`.
+  - If the **initialization inferred type** on one or more reachable paths is
+    the same type `T` (using syntactic equality), and the **initialization
+    inferred type** on all other reachable paths is `Null`, then the
+    **initialization inferred type** after the join is `T?`.
+  - Otherwise, it is an error.
```

Comment:
How about the situation where three branches has the types `int`, `int?` and `Null`?
Or just two branches having the types `int` and `int?`. Which can happen with consecutive joins like;
```dart
var x;
if (test1) {
  x = 1;
} else if (test2) {
  x = null;
} else {
  x = 0;
}
```
Here the join points would first combine `x=null` and `x=0` to `int?`, then fail to join that with `x=1`.

So, reasonable code which should probably be accepted, because otherwise we can't explain this rule to users.

Maybe:

> - Otherwise, if there exists a type *T* such that **initialization inferred type** on all paths are either the same as *T* or *T*? (using syntactic equality) or it is `Null`, then the **initialization inferred type** after the join is *T*?.

Reply:
Good point, I'll take a crack at making this work.  Definitely starts getting complicated here though.